reader_concurrency_semaphore.cc

```diff
@@ -720,7 +720,18 @@ uint64_t reader_concurrency_semaphore::get_serialize_limit() const {
     return _initial_resources.memory * _serialize_limit_multiplier();
 }
 
+uint64_t reader_concurrency_semaphore::get_kill_limit() const {
+    if (!_kill_limit_multiplier() || _kill_limit_multiplier() == std::numeric_limits<uint32_t>::max() || is_unlimited()) {
+        return std::numeric_limits<uint64_t>::max();
+    }
+    return _initial_resources.memory * _kill_limit_multiplier();
+}
+
 void reader_concurrency_semaphore::consume(resources r) {
+    if ((consumed_resources().memory + r.memory) >= get_kill_limit()) {
+        ++_stats.oom_kills;
+        throw std::bad_alloc();
+    }
```

Comment:
This is a bit heavy-handed (killing the last one rather than the biggest one) but I think it's okay.

1. serialization should have protected us from all but the worst
2. killing "a few" queries and just the bad one isn't much different, compared to the alternative of a crash

So I think this is a reasonable tradeoff.